File: ml-benchmark/synduce/terms/height.f
Inductive ArithOp = aPlus Unit | aMinus Unit | aGt Unit;
Inductive BoolOp = bNot Unit | bAnd Unit | bOr Unit | bEq Unit;
Inductive Term =
  tArithBin {ArithOp, Term, Term}
| tBoolBin {BoolOp, Term, Term}
| tArithUn {ArithOp, Term}
| tBoolUn {BoolOp, Term}
| tVar Int | tCInt Int | tCBool Bool;

Inductive Op = opPlus Unit | opMinus Unit | opNot Unit | opAnd Unit | opOr Unit | opGt Unit | opEq Unit;
Inductive Term2 =
  bin {Op, Term2, Term2}
| un {Op, Term2}
| var Int | cInt Int | cBool Bool;

mk_bin = \t1: Term. \t2: Term. \op: Op.
match op with
  opPlus _ -> tArithBin {aPlus unit, t1, t2}
| opMinus _ -> tArithBin {aMinus unit, t1, t2}
| opNot _ -> tBoolBin {bNot unit, t1, t2}
| opAnd _ -> tBoolBin {bAnd unit, t1, t2}
| opOr _ -> tBoolBin {bOr unit, t1, t2}
| opGt _ -> tArithBin {aGt unit, t1, t2}
| opEq _ -> tBoolBin {bEq unit, t1, t2}
end;

mk_un = \t: Term. \op: Op.
match op with
  opPlus _ -> tArithUn {aPlus unit, t}
| opMinus _ -> tArithUn {aMinus unit, t}
| opNot _ -> tBoolUn {bNot unit, t}
| opAnd _ -> tBoolUn {bAnd unit, t}
| opOr _ -> tBoolUn {bOr unit, t}
| opGt _ -> tArithUn {aGt unit, t}
| opEq _ -> tBoolUn {bEq unit, t}
end;

repr = fix (
  \f: Term2 -> Term. \t: Term2.
  match t with
    bin {op, t1, t2} -> mk_bin (f t1) (f t2) op
  | un {o, x} -> mk_un (f x) o
  | var i -> tVar i
  | cInt i -> tCInt i
  | cBool b -> tCBool b
  end
);

max = \a: Int. \b: Int. if < a b then b else a;

spec = fix (
  \f: Term -> Int. \t: Term.
  match t with
    tArithBin {op, t1, t2} -> + 1 (max (f t1) (f t2))
  | tBoolBin {op, t1, t2} -> + 1 (max (f t1) (f t2))
  | tArithUn {op, t} -> + 1 (f t)
  | tBoolUn {op, t} -> + 1 (f t)
  | tVar _ -> 1
  | tCInt _ -> 1
  | tCBool _ -> 1
  end
);

target = fix (
  \f: Term2 -> Compress Term2. \t: Term2.
  match t with
    bin {op, t1, t2} -> bin {op, f t1, f t2}
  | un {op, t} -> un {op, f t}
  | var i -> var i
  | cInt i -> cInt i
  | cBool i -> cBool i
  end
);

main = \xs: Term2. spec (repr (target xs));

File: ml-benchmark/synduce/terms/height.ml
type arithOp = APlus of unit | AMinus of unit | AGt of unit
type boolOp = BNot of unit | BAnd of unit | BOr of unit | BEq of unit
type term =
  | TArithBin of arithOp * term * term
  | TBoolBin of boolOp * term * term
  | TArithUn of arithOp * term
  | TBoolUn of boolOp * term
  | TVar of int
  | TCInt of int
  | TCBool of bool

type op = OpPlus of unit | OpMinus of unit | OpNot of unit | OpAnd of unit | OpOr of unit | OpGt of unit | OpEq of unit
type term2 = Bin of op * term2 * term2 | Un of op * term2 | Var of int | CInt of int | CBool of bool

let mk_bin t1 t2 op =
  match op with
  | OpPlus _ -> TArithBin (APlus (), t1, t2)
  | OpMinus _ -> TArithBin (AMinus (), t1, t2)
  | OpNot _ -> TBoolBin (BNot (), t1, t2)
  | OpAnd _ -> TBoolBin (BAnd (), t1, t2)
  | OpOr _ -> TBoolBin (BOr (), t1, t2)
  | OpGt _ -> TArithBin (AGt (), t1, t2)
  | OpEq _ -> TBoolBin (BEq (), t1, t2)

let mk_un t op =
  match op with
  | OpPlus _ -> TArithUn (APlus (), t)
  | OpMinus _ -> TArithUn (AMinus (), t)
  | OpNot _ -> TBoolUn (BNot (), t)
  | OpAnd _ -> TBoolUn (BAnd (), t)
  | OpOr _ -> TBoolUn (BOr (), t)
  | OpGt _ -> TArithUn (AGt (), t)
  | OpEq _ -> TBoolUn (BEq (), t)

let rec repr t =
  match t with
  | Bin (op, t1, t2) -> mk_bin (repr t1) (repr t2) op
  | Un (o, x) -> mk_un (repr x) o
  | Var i -> TVar i
  | CInt i -> TCInt i
  | CBool b -> TCBool b

let max a b =
  if a < b then b else a

let rec spec t =
  match t with
  | TArithBin (op, t1, t2) -> 1 + (max (spec t1) (spec t2))
  | TBoolBin (op, t1, t2) -> 1 + (max (spec t1) (spec t2))
  | TArithUn (op, t) -> 1 + (spec t)
  | TBoolUn (op, t) -> 1 + (spec t)
  | TVar _ -> 1
  | TCInt _ -> 1
  | TCBool _ -> 1

let rec target t =
  match t with
  | Bin (op, t1, t2) -> Bin (op, target t1, target t2)
  | Un (op, t) -> Un (op, target t)
  | Var i -> Var i
  | CInt i -> CInt i
  | CBool i -> CBool i

let program xs = spec (repr (target xs))